ffi/src/lib.rs

```diff
@@ -313,162 +568,365 @@ fn unwrap_conversion<T>(value: Option<T>, expected_type: ValueType) -> T {
+/// If the value of the boolean is `false` the value returned is 0.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::Boolean`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_as_boolean(typed_value: *mut TypedValue) -> i32 {
     let typed_value = Box::from_raw(typed_value);
     if unwrap_conversion(typed_value.into_boolean(), ValueType::Boolean) { 1 } else { 0 }
 }
 
-//as_double
+/// Consumes a `TypedValue` and returns the value as a `f64`.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::Double`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_as_double(typed_value: *mut TypedValue) ->  f64 {
     let typed_value = Box::from_raw(typed_value);
     unwrap_conversion(typed_value.into_double(), ValueType::Double)
 }
 
-//as_timestamp
+/// Consumes a `TypedValue` and returns the value as a microsecond timestamp.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::Instant`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_as_timestamp(typed_value: *mut TypedValue) ->  c_longlong {
     let typed_value = Box::from_raw(typed_value);
     unwrap_conversion(typed_value.into_timestamp(), ValueType::Instant)
 }
 
-//as_string
+/// Consumes a `TypedValue` and returns the value as a `String`.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::String`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_as_string(typed_value: *mut TypedValue) ->  *const c_char {
     let typed_value = Box::from_raw(typed_value);
     c_char_from_rc(unwrap_conversion(typed_value.into_string(), ValueType::String))
 }
 
-//as_uuid
+/// Consumes a `TypedValue` and returns the value as a UUID byte slice of length 16.
+///
+/// # Panics
+///
+/// If the `ValueType` of the `TypedValue` is not `ValueType::Uuid`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_as_uuid(typed_value: *mut TypedValue) ->  *mut [u8; 16] {
     let typed_value = Box::from_raw(typed_value);
     let value = unwrap_conversion(typed_value.into_uuid(), ValueType::Uuid);
     Box::into_raw(Box::new(*value.as_bytes()))
 }
 
-//value_type
+/// Returns the `ValueType` of this `TypedValue`.
+///
+/// # Safety
+///
+/// Callers must ensure that the pointer to the `TypedValue` is not dangling .
 #[no_mangle]
 pub unsafe extern "C" fn typed_value_value_type(typed_value: *mut TypedValue) ->  ValueType {
     let typed_value = &*typed_value;
```

Comment:
Is this mangling (`&*`) necessary?  That's syntactically a no-op, so I'm surprised the compiler needs it.

nit: there seem to be extra spaces before the return types `->  ValueType`, perhaps throughout.  Copy-paste error?

Reply:
Yes, the `&*` is necessary. Dereference the raw pointer (`*mut TypedValue`) and then create a rust reference to it. This is the standard way of referencing an raw pointer directly without reclaiming ownership.

Reply:
Thanks!  TIL.